pyvptree/include/VPTree.hpp

```diff
@@ -267,37 +147,30 @@ template <typename T, typename distance_type, distance_type (*distance)(const T
             throw new std::invalid_argument("invalid state - checksum mismatch");
         }
 
-        uint8_t *p_buffer = &const_cast<std::vector<uint8_t> &>(state.data)[0];
-        size_t elem_size = *((size_t *)p_buffer);
-        p_buffer += sizeof(size_t);
+        SerializedState copy(state);
 
-        size_t num_elements_per_example = *((size_t *)p_buffer);
-        p_buffer += sizeof(size_t);
+        _rootPartition = new VPLevelPartition<distance_type>();
 
-        size_t num_examples = *((size_t *)p_buffer);
-        p_buffer += sizeof(size_t);
+        size_t elem_size = copy.pop<size_t>();
+        size_t num_elements_per_example = copy.pop<size_t>();
+        size_t num_examples = copy.pop<size_t>();
 
-        _examples.clear();
         _examples.reserve(num_examples);
         _examples.resize(num_examples);
-        for (size_t i = 0; i < num_examples; ++i) {
-            unsigned int originalIndex = *((unsigned int *)(p_buffer));
-            p_buffer += sizeof(unsigned int);
+        for (int64_t i = num_examples - 1; i >= 0; --i) {
 
             auto &example = _examples[i];
-            example.originalIndex = originalIndex;
             auto &val = example.val;
             val.resize(num_elements_per_example);
 
-            for (size_t j = 0; j < num_elements_per_example; ++j) {
-                std::memcpy(&val[j], p_buffer, elem_size);
-                p_buffer += elem_size;
+            for (int64_t j = num_elements_per_example - 1; j >= 0; --j) {
+                copy.pop_by_size(&val[j], elem_size);
             }
+            int64_t originalIndex = copy.pop<int64_t>();
+            example.originalIndex = originalIndex;
         }
 
-        _rootPartition = new VPLevelPartition<distance_type>();
-        uint32_t offset = p_buffer - &state.data[0];
-        _rootPartition->deserialize(state, offset);
+        _rootPartition->deserialize(copy);
     }
 
     void searchKNN(const std::vector<T> &queries, unsigned int k, std::vector<VPTreeSearchResultElement> &results) {
```

Comment:
should it be `size_t k`?

Reply:
I think for `k` here is the number of neighbors, should be ok to use unsigned int?

Reply:
Oh, sorry, now I understand what you meant. K is ate most the number of elements, so yes, indeed you are right, its better to be of type `size_t`. Will fix it. Thanks.

Reply:
Fixed here: d6959b1152ff406b51a364cebfe078a295c04ccd